torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -427,21 +427,30 @@ bool ProcessGroupNCCL::WorkNCCL::startedGPUExecutionInternal() const {
 }
 
 bool ProcessGroupNCCL::WorkNCCL::finishedGPUExecutionInternal() const {
+  auto mode = cudaStreamCaptureModeThreadLocal;
+  // push
+  cudaThreadExchangeStreamCaptureMode(&mode);
   try {
     for (const auto i : c10::irange(devices_.size())) {
       // Checking the work's corresponding CUDA events' status
       if (!(*ncclEndEvents_)[i].query()) {
+        // pop
+        cudaThreadExchangeStreamCaptureMode(&mode);
         return false;
       }
     }
   } catch (const std::exception& e) {
+    // pop
+    cudaThreadExchangeStreamCaptureMode(&mode);
```

Comment:
If this catch branch runs and doesn't re-throw, don't we end up double-popping (aka leaving it as 'local') when `finishedGPUExecutionInternal` finishes?